File: server/api/presenters/paper_trading.py
from __future__ import annotations

from typing import Any

from server.api.presenters.common import table_payload, to_browser_value


def present_snapshot(value: dict[str, Any]) -> dict[str, Any]:
    data = dict(value or {})
    return {
        "user_id": str(data.get("user_id") or ""),
        "is_available": bool(data.get("is_available")),
        "account": to_browser_value(data.get("account") or {}),
        "positions": table_payload(data.get("positions")),
        "orders": table_payload(data.get("orders")),
        "nav_history": table_payload(data.get("nav_history")),
        "decisions": table_payload(data.get("decisions") or []),
        "risk_report": to_browser_value(data.get("risk_report") or {}),
        "execution_diagnostics": to_browser_value(data.get("execution_diagnostics") or {}),
        "trading_settings": to_browser_value(data.get("trading_settings") or {}),
        "order_snapshot_dates": to_browser_value(data.get("order_snapshot_dates") or []),
        "position_snapshot_dates": to_browser_value(data.get("position_snapshot_dates") or []),
        "profile": to_browser_value(data.get("profile") or {}),
        "profile_complete": bool(data.get("profile_complete")),
        "profile_options": to_browser_value(data.get("profile_options") or {}),
        "cash_flows": table_payload(data.get("cash_flows") or []),
        "backfill_status": to_browser_value(data.get("backfill_status") or {}),
        "ai_reliability": to_browser_value(data.get("ai_reliability") or {}),
        "scheduler": to_browser_value(data.get("scheduler") or {}),
    }


def present_daily_history(value: dict[str, Any]) -> dict[str, Any]:
    data = dict(value or {})
    return {
        "user_id": str(data.get("user_id") or ""),
        "trade_date": str(data.get("trade_date") or ""),
        "available_dates": to_browser_value(data.get("available_dates") or []),
        "has_position_snapshot": bool(data.get("has_position_snapshot")),
        "positions": table_payload(data.get("positions")),
        "operations": table_payload(data.get("operations")),
        "summary": to_browser_value(data.get("summary") or {}),
    }


def present_profile(value: dict[str, Any]) -> dict[str, Any]:
    data = dict(value or {})
    return {
        "user_id": str(data.get("user_id") or ""),
        "profile": to_browser_value(data.get("profile") or {}),
        "complete": bool(data.get("complete")),
        "options": to_browser_value(data.get("options") or {}),
    }


def present_proposals(value: list[dict[str, Any]]) -> dict[str, Any]:
    records = to_browser_value(value or [])
    if not isinstance(records, list):
        records = []
    return {"records": records, "total": len(records)}


def present_write_result(value: dict[str, Any]) -> dict[str, Any]:
    return to_browser_value(value or {})

File: frontend/src/types/paperTrading.ts
import type { TablePayload } from './common'

export type GenericRecord = Record<string, unknown>

export interface PaperTradingSnapshot {
  user_id: string
  is_available: boolean
  account: GenericRecord
  positions: TablePayload<GenericRecord>
  orders: TablePayload<GenericRecord>
  nav_history: TablePayload<GenericRecord>
  decisions: TablePayload<GenericRecord>
  risk_report: GenericRecord
  execution_diagnostics: GenericRecord
  trading_settings: GenericRecord
  order_snapshot_dates: unknown[]
  position_snapshot_dates: unknown[]
  profile: GenericRecord
  profile_complete: boolean
  profile_options: Record<string, string[]>
  cash_flows: TablePayload<GenericRecord>
  backfill_status: GenericRecord
  ai_reliability: GenericRecord
  scheduler: GenericRecord
}

export interface PaperTradingHistorySummary {
  position_count: number
  operation_count: number
  buy_count: number
  sell_count: number
  ohlc_matched_count: number
  ohlc_missing_count: number
}

export interface PaperTradingDayHistory {
  user_id: string
  trade_date: string
  available_dates: string[]
  has_position_snapshot: boolean
  positions: TablePayload<GenericRecord>
  operations: TablePayload<GenericRecord>
  summary: PaperTradingHistorySummary
}

export interface PaperProfilePayload {
  user_id: string
  profile: GenericRecord
  complete: boolean
  options: Record<string, string[]>
}

export interface PaperProposal {
  plan_id: string
  intent: string
  operation_type: string
  confirmation_status: string
  execution_status: string
  expires_at?: string | null
  created_at?: string | null
  before_state_summary: GenericRecord
  proposed_changes: unknown[]
  after_state_preview: GenericRecord
  warnings: unknown[]
  validation_results: GenericRecord
  confirmation_phrase: string
  token_present: boolean
}

export interface PaperProposalList {
  records: PaperProposal[]
  total: number
}

export interface WriteRequestMeta {
  request_id: string
  idempotency_key: string
}

export interface ProfileUpdateRequest extends WriteRequestMeta {
  user_id: string
  profile: GenericRecord
  confirmed: boolean
}

export interface CashFlowPreviewRequest extends WriteRequestMeta {
  user_id: string
  flow_type: 'deposit' | 'withdrawal'
  amount: number
  effective_date?: string
  reason?: string
}

export interface BackfillPreviewRequest extends WriteRequestMeta {
  user_id: string
  start_date: string
  end_date: string
  initial_cash?: number
  force: boolean
  resume: boolean
}

export interface ProposalCommitRequest extends WriteRequestMeta {
  user_id: string
  confirmation_text: string
}

export interface ProposalRejectRequest extends WriteRequestMeta {
  user_id: string
  reason?: string
}

export interface CashFlowCancelRequest extends WriteRequestMeta {
  user_id: string
  confirmed: boolean
}
